File: templates/problems/01_arrays_hashing/03_two_sum/typescript/solution.ts
/**
 * Find two indices whose values add to target using a hash map.
 * Time: O(n), Space: O(n)
 */
export function solveTwoSum(nums: number[], target: number): number[] {
    const seen = new Map<number, number>();
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if (seen.has(complement)) {
            return [seen.get(complement)!, i];
        }
        seen.set(nums[i], i);
    }
    return [];
}

File: templates/problems/01_arrays_hashing/03_two_sum/python/solution.py
def solve_two_sum(nums: list[int], target: int) -> list[int]:
    """Find two indices whose values add to target using a hash map.

    Time: O(n), Space: O(n)
    """
    seen = {}
    for i, n in enumerate(nums):
        complement = target - n
        if complement in seen:
            return [seen[complement], i]
        seen[n] = i
    return []
